src/StdChains.sol

```diff
@@ -20,70 +22,131 @@ abstract contract StdChains {
+    mapping(uint256 => string) private idToAlias;
 
-    function getChain(string memory key) internal virtual returns (Chain memory) {
+    // The RPC URL will be fetched from config or defaultRpcUrls if possible.
+    function getChain(string memory chainAlias) internal virtual returns (Chain memory chain) {
         initialize();
-        return chains[key];
+        chain = chains[chainAlias];
+        require(
+            chain.chainId != 0,
+            string(abi.encodePacked("StdChains getChain(string): Chain alias \"", chainAlias, "\" not found."))
+        );
+
+        withRpcUrl(chainAlias, chain);
     }
 
-    function getChain(uint256 chainId) internal virtual returns (Chain memory) {
+    function getChain(uint256 chainId) internal virtual returns (Chain memory chain) {
         initialize();
-        return chains[idToKey[chainId]];
+        string memory chainAlias = idToAlias[chainId];
+
+        require(
+            bytes(idToAlias[chainId]).length != 0,
+            string(
+                abi.encodePacked("StdChains getChain(uint): No alias found for chain ID ", vm.toString(chainId), ".")
+            )
+        );
+
+        chain = chains[chainAlias];
+
+        require(
+            chain.chainId != 0,
+            string(abi.encodePacked("StdChains getChain(uint): Chain ID ", vm.toString(chainId), " not found."))
+        );
+
+        withRpcUrl(chainAlias, chain);
+    }
+
+    // lookup rpcUrl, in descending order of priority:
+    // current -> config (foundry.toml) -> default
+    function withRpcUrl(string memory chainAlias, Chain memory chain) private view {
+        if (bytes(chain.rpcUrl).length == 0) {
+            try vm.rpcUrl(chainAlias) returns (string memory configRpcUrl) {
+                chain.rpcUrl = configRpcUrl;
+            } catch (bytes memory err) {
+                chain.rpcUrl = defaultRpcUrls[chainAlias];
+                // distinguish 'not found' from 'cannot read'
+                bytes memory notFoundError =
+                    abi.encodeWithSignature("CheatCodeError", string(abi.encodePacked("invalid rpc url ", chainAlias)));
+                if (keccak256(notFoundError) != keccak256(err) || bytes(chain.rpcUrl).length == 0) {
+                    /// @solidity memory-safe-assembly
+                    assembly {
+                        revert(add(32, err), mload(err))
+                    }
+                }
+            }
+        }
     }
 
-    function setChain(string memory key, string memory name, uint256 chainId, string memory rpcUrl) internal virtual {
+    // set chain info, with priority to chainAlias' rpc url in foundry.toml
+    function setChainWithDefaultRpcUrl(string memory chainAlias, Chain memory chain) internal {
+        string memory rpcUrl = chain.rpcUrl;
+        defaultRpcUrls[chainAlias] = rpcUrl;
+        chain.rpcUrl = "";
+        setChain(chainAlias, chain);
+        chain.rpcUrl = rpcUrl; // restore argument
```

Comment:
`chain` is a memory reference. All modifications to it will be visible throughout the current EVM call frame. Not restoring the `rpcUrl` means that `setChainWithDefaultRpcUrl` sets its argument's `rpcUrl` property to the empty.